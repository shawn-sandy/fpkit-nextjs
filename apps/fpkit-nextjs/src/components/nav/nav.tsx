/**
 * Nav component
 *
 * The Nav component is used to wrap a group of links that are related
 * to each other. The Nav component is used by the Header component.
 *
 * @param {NavProps} props
 * @returns {JSX.Element}
 */

import React, { Children } from "react";

import { Tag } from "@fpkit/react";

export type NavProps = {
  children?: React.ReactNode;
};

export const Nav = ({ children }: NavProps) => {
  return (
    <>
      <Tag as="nav">{children}</Tag>
    </>
  );
};

export default Nav;
Nav.displayName = "Nav";

/**
 * NavItem component
 * Creates a list item for the Nav component
 * @param param0
 * @returns
 */

export type NavItemProps = {
  children?: React.ReactNode;
};
export const NavItem = ({ children }: NavProps) => {
  return (
    <>
      <Tag as="li">{children}</Tag>
    </>
  );
};
