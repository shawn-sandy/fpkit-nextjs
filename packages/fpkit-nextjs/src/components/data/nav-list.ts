
export const navListItems = [
    {
        kind: "MdxPage",
        name: "articles",
        route: "/articles",
        frontMatter: {
            type: "page",
            title: "Articles",
            date: "2020-01-01T00:00:00.000Z",
        },
    },
    {
        kind: "MdxPage",
        name: "index",
        route: "/",
        frontMatter: {
            type: "page",
            title: "About",
            date: "2020-01-01T00:00:00.000Z",
        },
    },
    {
        kind: "MdxPage",
        name: "posts",
        route: "/posts",
        frontMatter: {
            type: "page",
            title: "Posts",
            date: "2020-01-01T00:00:00.000Z",
        },
    },
];
