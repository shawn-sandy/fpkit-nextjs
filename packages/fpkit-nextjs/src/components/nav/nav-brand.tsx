import React from "react";

import { Tag, Title } from "@fpkit/react";

export interface BrandTypes {
  url: string;
  logo: React.ReactNode;
}

export const NavBrand = ({ url, logo }: BrandTypes) => {
  return (
    <Tag as="div">
      <Title elm="h1" styles={{ "--fs": "1.2rem" }}>
        <Tag as="a" href={url} styles={{ textTransform: "capitalize" }}>
          {logo}
        </Tag>
      </Title>
    </Tag>
  );
};

export default NavBrand;
