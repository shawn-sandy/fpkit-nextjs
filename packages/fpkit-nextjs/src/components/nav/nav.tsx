import React from "react";

import { Tag } from "@fpkit/react";

export type NavItemProps = {
  route: string;
  styles?: {};
  children?: React.ReactNode;
};

/**
 * NavItem component creates a navigation link element wrapped in a list item `li`
 * @param param0
 * @returns
 */
export const NavItem = ({
  route,
  styles,
  children,
  ...props
}: NavItemProps) => {
  return (
    <>
      <Tag as="li" {...props}>
        <Tag as="a" href={route} styles={{ textTransform: "capitalize" }}>
          {children}
        </Tag>
      </Tag>
    </>
  );
};

export type NavProps = {
  children?: React.ReactNode;
};

/**
 * Nav component creates a navigation element
 * @param param0
 * @returns
 */
export const Nav = ({ children, ...props }: NavProps) => {
  return (
    <>
      <Tag as="nav" {...props}>
        {children}
      </Tag>
    </>
  );
};

/**
 * Exports
 */
export default Nav;
Nav.displayName = "Nav";
