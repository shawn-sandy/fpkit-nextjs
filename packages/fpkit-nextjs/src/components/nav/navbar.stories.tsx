import { StoryObj, Meta } from "@storybook/react";

import { within } from "@storybook/testing-library";
import { expect } from "@storybook/jest";

import Navbar from "./navbar";

const meta: Meta<typeof Navbar> = {
  title: "Nextra Components/Navbar",
  component: Navbar,
};

export default meta;
type Story = StoryObj<typeof Navbar>;

const navList = [
  {
    route: "/home",
    name: "Home",
  },
  {
    route: "/about",
    name: "About",
  },
  {
    route: "/contact",
    name: "Contact",
  },
];

export const BaseNavbar: Story = {
  args: {
    navList: navList,
  },
  play: async ({ canvasElement }) => {
    const canvas = within(canvasElement);
    const nav = canvas.getByRole("navigation");
    expect(nav).toBeDefined();
    const links = canvas.getAllByRole("link");
    expect(links).toHaveLength(navList.length);
  },
};
