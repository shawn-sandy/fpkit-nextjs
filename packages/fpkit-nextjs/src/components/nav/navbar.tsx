import React from "react";
import { Tag, Text, Title } from "@fpkit/react";
import Nav, { NavItem } from "./nav";
import { PageMapItem } from "nextra";

export type NavbarProps = {
  route: string;
  name: string;
};

export type NavListTypes = {
  navList: NavbarProps[];
  brand?: {
    url?: string;
    logo?: React.ReactNode;
  };
};

export const Navbar = ({ navList, brand }: NavListTypes) => {
  return (
    <Nav>
      <Tag as="div">
        {!!brand && (
          <Title elm="h1" styles={{ "--fs": "1.2rem" }}>
            <Tag
              as="a"
              href={brand.url}
              styles={{ textTransform: "capitalize" }}
            >
              {brand.logo}
            </Tag>
          </Title>
        )}
      </Tag>
      <Tag as="ul">
        {navList.map((navItem) => {
          return (
            <NavItem route={navItem.route} key={React.useId()}>
              {navItem.name}
            </NavItem>
          );
        })}
      </Tag>
    </Nav>
  );
};

export default Navbar;
Navbar.displayName = "Navbar";
