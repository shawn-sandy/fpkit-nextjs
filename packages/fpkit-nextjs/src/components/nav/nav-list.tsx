import React from "react";
import { NavItem } from "./nav";

import { FrontMatter } from "nextra";
import { Tag } from "@fpkit/react";

export interface Page {
  kind: string;
  name: string;
  route: string;
  frontMatter: FrontMatter;
}

export interface NavListProps {
  pageList: Page[];
}

export const NavList = ({ pageList }: NavListProps) => {
  // throw an error if the pageList is empty or undefined
  if (!pageList) {
    throw new Error("pageList is empty or undefined");
  }
  return (
    <Tag as="ul">
      {pageList.map((page, index) => (
        <NavItem route={page.route} key={index}>
          {page.frontMatter.title}
        </NavItem>
      ))}
    </Tag>
  );
};

export default NavList;
