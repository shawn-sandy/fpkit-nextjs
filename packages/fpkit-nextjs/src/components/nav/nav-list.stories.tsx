import { StoryObj, Meta } from "@storybook/react";

import { within } from "@storybook/testing-library";
import { expect } from "@storybook/jest";
import { Nav } from "./nav";

import NavList from "./nav-list";

import { navListItems } from "../data/nav-list";

const meta: Meta<typeof NavList> = {
  title: "Nextra Components/Navbar",
  component: NavList,
  // @ts-ignore
  //   subcomponents: { NavItem },
  args: {},
};

export default meta;
type Story = StoryObj<typeof NavList>;

export const List: Story = {
  args: {
    pageList: navListItems,
  },
  play: async ({ canvasElement }) => {
    const canvas = within(canvasElement);
    const list = canvas.getByRole("list");
    const links = canvas.getAllByRole("link");

    // @ts-ignore
    expect(list).toBeInTheDocument();
    expect(links).toHaveLength(navListItems.length);
  },
};

export const RenderNavbar: Story = {
  args: {
    ...List.args,
  },
  render: ({ ...args }) => (
    <Nav>
      <NavList {...args} />
    </Nav>
  ),
  play: async ({ canvasElement }) => {
    const canvas = within(canvasElement);
    // @ts-ignore
    await Default.play({ canvasElement });

    const nav = canvas.getByRole("navigation");
    // @ts-ignore
    expect(nav).toBeInTheDocument();
  },
};
