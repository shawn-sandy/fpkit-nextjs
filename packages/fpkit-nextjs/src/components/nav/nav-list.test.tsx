import React from "react";

import { describe, expect, test } from "vitest";
import { render, screen } from "@testing-library/react";
import { axe } from "vitest-axe";

// ðŸ‘‰ import ComponentName from "./component-name";
import { NavList } from "./nav-list";
import { Nav } from "./nav";

import { navListItems } from "../data/nav-list";

describe("Nav renders correctly", () => {
  test("renders without crashing", async () => {
    // ðŸ‘‰ render(<ComponentName>Link</ComponentName>);
    render(
      <Nav>
        <NavList pageList={navListItems} />
      </Nav>
    );
    const nav = screen.getByRole("navigation");
    const list = screen.getByRole("list");
    const links = screen.getAllByRole("link");

    expect(nav).toBeDefined();
    expect(list).toBeDefined();
    expect(links).toHaveLength(navListItems.length);
    expect(await axe(nav)).toHaveNoViolations();
    expect(nav).toMatchSnapshot();
  });
});
