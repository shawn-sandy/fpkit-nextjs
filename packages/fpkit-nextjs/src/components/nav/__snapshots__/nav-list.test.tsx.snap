// Vitest Snapshot v1

exports[`Nav renders correctly > renders without crashing 1`] = `
<nav>
  <ul>
    <li>
      <a
        href="/articles"
        style="text-transform: capitalize;"
      >
        Articles
      </a>
    </li>
    <li>
      <a
        href="/"
        style="text-transform: capitalize;"
      >
        About
      </a>
    </li>
    <li>
      <a
        href="/posts"
        style="text-transform: capitalize;"
      >
        Posts
      </a>
    </li>
  </ul>
</nav>
`;
